src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1619,21 +1619,41 @@ private void NotifyInvoker(string functionName, Exception ex)
 
         private void OnFileChanged(FileSystemEventArgs e)
         {
-            string directory = GetRelativeDirectory(e.FullPath, ScriptConfig.RootScriptPath);
-            bool isWatchedDirectory = ScriptConfig.WatchDirectories.Contains(directory);
-
             // We will perform a host restart in the following cases:
             // - the file change was under one of the configured watched directories (e.g. node_modules, shared code directories, etc.)
             // - the host.json file was changed
             // - a function.json file was changed
+            // - a proxies.json file was changed
             // - a function directory was added/removed/renamed
             // A full host shutdown is performed when an assembly (.dll, .exe) in a watched directory is modified
+
+            string changeDesciprtioin = string.Empty;
+            string directory = GetRelativeDirectory(e.FullPath, ScriptConfig.RootScriptPath);
             string fileName = Path.GetFileName(e.Name);
-            if (isWatchedDirectory ||
-                (string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0 ||
-                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0 ||
-                 string.Compare(fileName, ScriptConstants.ProxyMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
-                !_directorySnapshot.SequenceEqual(Directory.EnumerateDirectories(ScriptConfig.RootScriptPath)))
+
+            if (ScriptConfig.WatchDirectories.Contains(directory) ||
+                string.Equals("bin", directory, StringComparison.OrdinalIgnoreCase))
+            {
+                changeDesciprtioin = "Watched directory";
+            }
+            else
+            {
+                if (string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0 ||
+                    string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0 ||
+                    string.Compare(fileName, ScriptConstants.ProxyMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0)
+                {
+                    changeDesciprtioin = "File";
+                }
+                else
+                {
+                    if (!_directorySnapshot.SequenceEqual(Directory.EnumerateDirectories(ScriptConfig.RootScriptPath)))
```

Comment:
One of the things the issue mentions is only performing the directory enumeration and snapshot checks only if the path is a directory, which, in addition to prevent this from happening a lot more often than it should, would also prevent false positives that ended up driving changes like what is described in #1903 . With that in place, we can also address and close #1903